### ISSUE TITLE
eslint location list results don't jump to line

### ISSUE THREAD

**Issue by mqudsi** [neomake/neomake#1582]
Using the default eslint maker (`Neomake eslint`) with Neomake 8794ee654a715867d34045af49dd00e8474b0850 under neovim v0.2.1-855-gcb912a3 results in the linter correctly catching errors and their line numbers being displayed in the the location list, but attempting to jump to an error does nothing.

What shows up in the location list:

```
cpsatv.js|1 col 1 warning| Definition for rule 'react/jsx-sort-prop-types' was not found (react/jsx-sort-prop-types)
cpsatv.js|1 col 1 warning| Definition for rule 'react/require-extension' was not found (react/require-extension)
cpsatv.js|1 col 1 warning| Definition for rule 'react/wrap-multilines' was not found (react/wrap-multilines)
cpsatv.js|38 col 58 error| Expected '===' and instead saw '=='. (eqeqeq)
cpsatv.js|64 col 12 warning| Gratuitous parentheses around expression. (no-extra-parens)
cpsatv.js|64 col 19 error| Expected '!==' and instead saw '!='. (eqeqeq)
cpsatv.js|64 col 40 error| Expected '!==' and instead saw '!='. (eqeqeq)
cpsatv.js|102 col 17 warning| Gratuitous parentheses around expression. (no-extra-parens)
cpsatv.js|102 col 35 error| Expected '!==' and instead saw '!='. (eqeqeq)
cpsatv.js|102 col 53 warning| Gratuitous parentheses around expression. (no-extra-parens)
cpsatv.js|105 col 17 warning| Gratuitous parentheses around expression. (no-extra-parens)
cpsatv.js|105 col 36 error| Expected '!==' and instead saw '!='. (eqeqeq)
cpsatv.js|105 col 54 warning| Gratuitous parentheses around expression. (no-extra-parens)
cpsatv.js|108 col 25 error| Expected '!==' and instead saw '!='. (eqeqeq)
cpsatv.js|144 col 34 error| Expected '!==' and instead saw '!='. (eqeqeq)
cpsatv.js|169 col 98 warning| Missing semicolon. (semi)
cpsatv.js|177 col 19 error| Expected '!==' and instead saw '!='. (eqeqeq)
cpsatv.js|215 col 22 error| Expected '===' and instead saw '=='. (eqeqeq)
cpsatv.js|218 col 12 error| Expected '!==' and instead saw '!='. (eqeqeq)
cpsatv.js|222 col 2 error| Mixed spaces and tabs. (no-mixed-spaces-and-tabs)
cpsatv.js|223 col 2 error| Mixed spaces and tabs. (no-mixed-spaces-and-tabs)
cpsatv.js|227 col 27 error| Expected '===' and instead saw '=='. (eqeqeq)
cpsatv.js|230 col 27 error| Expected '===' and instead saw '=='. (eqeqeq)
cpsatv.js|232 col 40 warning| Missing semicolon. (semi)
cpsatv.js|236 col 27 error| Expected '===' and instead saw '=='. (eqeqeq)
cpsatv.js|237 col 7 error| 'div' is already defined. (no-redeclare)
cpsatv.js|240 col 27 error| Expected '===' and instead saw '=='. (eqeqeq)
cpsatv.js|244 col 4 warning| Missing semicolon. (semi)
cpsatv.js|249 col 24 error| Expected '===' and instead saw '=='. (eqeqeq)
cpsatv.js|251 col 7 error| 'lines' is already defined. (no-redeclare)
cpsatv.js|255 col 7 error| 'lines' is already defined. (no-redeclare)
```

The first three fake results (they aren't an error with the actual file, but general warnings from eslint) jump to location 1:1 fine, but from the 3rd error onwards, nothing happens.

The NeomakeInfo output is below:

#### Neomake debug information
	Async support: 1
	Current filetype: javascript
	Windows: 0
	[shell, shellcmdflag, shellslash]: ['/bin/bash', '-c', 0]

##### Enabled makers
	For the current filetype ("javascript", used with :Neomake):
	 - jshint
	   - append_file: 1
	   - args: ['--verbose']
	   - auto_enabled: 1
	   - errorformat: '%A%f: line %l\, col %v\, %m \(%t%*\d\)'
	   - exe: 'jshint'
	   - postprocess: function('neomake#postprocess#GenericLengthPostprocess')
	   - WARNINGS:
		 - maker's exe (jshint) is not executable.
	 - jscs
	   - append_file: 1
	   - args: ['--no-colors', '--reporter', 'inline']
	   - auto_enabled: 1
	   - errorformat: '%E%f: line %l\, col %c\, %m'
	   - exe: 'jscs'
	 - eslint
	   - append_file: 1
	   - args: ['-f', 'compact']
	   - auto_enabled: 1
	   - errorformat: '%E%f: line %l\, col %c\, Error - %m,%W%f: line %l\, col %c\, Warning - %m,%-G,%-G%*\d problems%#'
	   - exe: 'eslint'
	NOTE: you can define g:neomake_javascript_enabled_makers to configure it (or b:neomake_javascript_enabled_makers).

	For the project (used with :Neomake!):
	 - makeprg
	   - args: []
	   - auto_enabled: 0
	   - errorformat: '%*[^"]"%f"%*\D%l: %m,"%f"%*\D%l: %m,%-G%f:%l: (Each undeclared identifier is reported only once,%-G%f:%l: for each function it appears in.),%-GIn file included from %f:%l:%c:,%-GIn file included from %f:%l:%c\,,%-GIn file included from %f:%l:%c,%-GIn file included from %f:%l,%-G%*[ ]from %f:%l:%c,%-G%*[ ]from %f:%l:,%-G%*[ ]from %f:%l\,,%-G%*[ ]from %f:%l,%f:%l:%c:%m,%f(%l):%m,%f:%l:%m,"%f"\, line %l%*\D%c%*[^ ] %m,%D%*\a[%*\d]: Entering directory %*[`'']%f'',%X%*\a[%*\d]: Leaving directory %*[`'']%f'',%D%*\a: Entering directory %*[`'']%f'',%X%*\a: Leaving directory %*[`'']%f'',%DMaking %*\a in %f,%f|%l| %m'
	   - exe: 'makeprg'
	   - fn: function('7')
	   - ERRORS:
		 - maker's exe (makeprg) is not execut
	Type  :quit<Enter>  to exit Nvim

**Comment by blueyed** [comment]
The debug log is missing..

What does `:echom string(getloclist(0)[0:10])` display?

**Comment by mqudsi** [comment]
There's something odder than this going on, depending on the timing of my interaction with the location list the jump sometimes works (and keeps working until Neomake is run again) or doesn't (and keeps not working until Neomake is run again). Except for that first 1:1 which always works.

**Using Neomake with both jshint and eslint installed:**

Additionally, if I interact with the location list immediately after Neomake runs and try to jump to an error, I get an error message `E92: Buffer 1634301033 not found` or similar, sometimes combined with "location list has changed". But selecting the same location list entry again without re-running Neomake or closing and reopening the location list magically works.

In trying to figure this out, I realized the the second error I'm reporting happens when I get a "Press ENTER or type command to continue" prompt in the vim command bar, which is due to a change in neovim or neomake that made my ":echom Build started" autohook sometimes require that "ENTER" for Neomake to begin:

![image](https://user-images.githubusercontent.com/606923/30453904-220dacfa-9960-11e7-8cb7-2ccd5f04eaca.png)

My hooks:

```vim
augroup neomake_hooks
    au!
    autocmd User NeomakeJobInit :call StartSpinner()
    autocmd User NeomakeJobInit :echom "Build started"
    autocmd User NeomakeFinished :call StopSpinner()
    autocmd User NeomakeFinished :echom "Build complete"
augroup END
```

I presume what's happening is that Neomake is triggered from the main buffer, taking its contents and associated configuration for the makers. But when I press "ENTER" in the command bar, Neomake takes the buffer id of that temporary message buffer as the buffer id for the jump lists, and when I press "ENTER" that buffer closes (hence the buffer not found message).

**Never mind**

I just disabled that hook, there was no "presss ENTER" message, but still the same error persisted.

The "buffer not found" message is always the same buffer ID  (and the buffer id doesn't look any more interesting in hexadecimal: 0x61697469), which is probably a clue.

**After removing eslint and using only jshint, the buffer not found message disappeared.** 

At this point, the previous behavior in my initial post about location list results not doing anything is what I observe. 

This is what `:echom string(getloclist(0)[0:1])` returns (where all entries are not working):

```
[{'lnum': 38, 'bufnr': 1, 'col': 37, 'valid': 1, 'vcol': 1, 'nr': -1, 'type': 'W', 'pattern': '', 'text': 'Functions declared within loops referencing an outer scoped variable may lead to c
onfusing semantics.'}, {'lnum': 153, 'bufnr': 1, 'col': 9, 'valid': 1, 'vcol': 1, 'nr': -1, 'type': 'W', 'pattern': '', 'text': 'Misleading line break before ''?''; readers may interpret th
is as an expression boundary.'}]
```

Which seems correct...

**Comment by mqudsi** [comment]
Wait, there may be something.

For this location list:

![image](https://user-images.githubusercontent.com/606923/30454529-435c79de-9962-11e7-8baa-d3f5942f4450.png)

It's not that it's not jumping to the location specified, it seems that some entries show one number but jump to another (the previous number). So when I'm already on that line, it appears as if nothing is happening.

For the above shown location list, 

``` json
[{'lnum': 38, 'bufnr': 1, 'col': 37, 'valid': 1, 'vcol': 1, 'nr': -1, 'type': 'W', 'pattern': '', 'text': 'Functions declared within loops referencing an outer scoped variable may lead to c
onfusing semantics.'}, {'lnum': 153, 'bufnr': 1, 'col': 9, 'valid': 1, 'vcol': 1, 'nr': -1, 'type': 'W', 'pattern': '', 'text': 'Misleading line break before ''?''; readers may interpret th
is as an expression boundary.'}, {'lnum': 223, 'bufnr': 1, 'col': 9, 'valid': 1, 'vcol': 1, 'nr': -1, 'type': 'W', 'pattern': '', 'text': 'It''s not necessary to initialize ''lines'' to ''u
ndefined''.'}, {'lnum': 224, 'bufnr': 1, 'col': 9, 'valid': 1, 'vcol': 1, 'nr': -1, 'type': 'W', 'pattern': '', 'text': 'It''s not necessary to initialize ''div'' to ''undefined''.'}, {'lnu
m': 0, 'bufnr': 0, 'col': 0, 'valid': 0, 'vcol': 0, 'nr': -1, 'type': 'W', 'pattern': '', 'text': ''}, {'lnum': 0, 'bufnr': 0, 'col': 0, 'valid': 0, 'vcol': 0, 'nr': -1, 'type': 'W', 'patte
rn': '', 'text': '4 errors'}, {'lnum': 0, 'bufnr': 0, 'col': 0, 'valid': 0, 'vcol': 0, 'nr': -1, 'type': 'W', 'pattern': '', 'text': 'No configuration found. Add a .jscsrc file to your proj
ect root or use the -c option.'}]
```

Yet the first entry takes me to line 38, the 2nd entry takes me to line 38, the 3rd entry takes me to line 153, and the 4th entry takes me to line 223.

It seems there's a ghost entry between the 1st and 2nd.

**Comment by mqudsi** [comment]
I'm not sure what to make of this, exporting the location list and re-importing it in a new session (copy and paste of getloclist/setloclist) has no problem navigating between those entries.

**Comment by blueyed** [comment]
Weird indeed.  Maybe a (Neo)vim bug after all?
Do you use vim-qf or other plugins interacting with the location list?

Does `:lfirst`, `:lnext` and `:lprev` behave better?

**Comment by blueyed** [comment]
Any new findings?

**Comment by mqudsi** [comment]
So I just ran into this again - it _may_ be a .js issue as that's what I'm working on once more today (and I rarely work on js). 

I have two errors in the location list, but double-clicking/entering on both goes nowhere. Interestingly, `:lfirst` runs (but does nothing), but then typing in `:lnext` gives an error about no more errors (when there should be a second (final) error.

```vim
:echom string(getloclist(0)[0:1])
```

gives the following (reformatted):

```vim
[{
    'lnum': 3,
    'bufnr': 1,
    'col': 25,
    'valid': 1,
    'vcol': 1,
    'nr': -1,
    'type': 'W',
    'pattern': '',
    'text': 'Missing semicolon.'
}, {
    'lnum': 90,
    'bufnr': 1,
    'col': 2,
    'valid': 1,
    'vcol': 1,
    'nr': -1,
    'type': 'W',
    'pattern': '',
    'text': 'Missing semicolon.'
}]
```

While I know used vim-qf at one point, I just checked and I'm definitely not using it now. My vim configuration is public at https://github.com/mqudsi/nvim-config but I don't think I have anything else. 

I do have [this script](https://github.com/mqudsi/nvim-config/blob/master/bufferclose.vim)  which exits vim if the only window left is the location list, but I just tried disabling it and could still reproduce the issue.

**Comment by blueyed** [comment]
Is the bufnr correct? (i.e. is it the second number in the last segment of your statusline - the `1` in `31:1`)
Can you provide the debug log when this happens? (see the "new issue template")

**Comment by joaosa** [comment]
I'm also experiencing this issue. @mqudsi did you manage to figure this out? :)

**Comment by blueyed** [comment]
Sounds really like a Vim bug, and there have been fixes made there, and also Neomake should behave/work around them.
If this still happens with latest Neomake for you, please re-open and provide more, or create a new issue.
There are several tests in Neomake for this.

**Comment by mqudsi** [comment]
I agree that the behavior observed (especially after dumping the contents of the quickfix list and observing that they are correct) is rather likely to be a (neo)vim big; there is no indication that neomake did anything incorrect here. 

What is odd is that I never experience this with any other method of populating the qflist.

**Comment by blueyed** [comment]
There are different methods to populate the list (functions (setqflist/setloclist) or commands (`:caddexpr`/`:laddexpr`)), which might have different behavior.
See e.g. https://github.com/vim/vim/issues/3633, where the workaround is to create a new list first and then append to it etc.
Might be useful to have a way to reproduce this, and I could look into this, but I suggest trying it with the latest Vim also.

**Comment by blueyed** [comment]
What would be important also is the neomake log, since it might also depend on if the list gets populated once, or multiple times.

You also mentioned:

> Additionally, if I interact with the location list immediately after Neomake runs and try to jump to an error, I get an error message E92: Buffer 1634301033 not found or similar, sometimes combined with "location list has changed". 

The first error is very odd - do you still see this?
As for the second I've also seen this rarely.  I assume it is related to using the list right after the hit-Enter prompt already ?  Maybe there is some queued Neomake action not handled correctly - here the log (and a new issue) would be required for debugging this.
